### DISCUSSION TITLE
Nuxt 3 Memory Leak

### DISCUSSION THREAD

**Discussion by LeonardoMussato** [nuxt/nuxt#16031] [Questions] [answered]
Hi, I'm trying to use nuxt 3 in a container on ECS (Amazon Elastic Container Service). It's not generating any error when it's built but it's generating a memory leak clearly visible from the logs of the container. Here are the logs:

![memoryLeak](https://user-images.githubusercontent.com/64805332/141837692-fdb9d603-932b-47b3-b133-51f01c128e53.png)
 
I tried to find any solution online but I wasn't lucky... Can you please tell me how to fix it or what can be a possible problem that can cause that?

Thank you for your help.

**Comment by pi0** [comment, marked as answer]
Hi @LeonardoMussato. Can you please somehow share your project?

**Reply by LeonardoMussato** [reply]
Sure here you can find the most updated version of the app based on nuxt3. The only thing that changes form the actual project is the url in the env files. 

Here is the link: https://github.com/LeonardoMussato/nuxt-memory-leak

Thank you so much for your time!

**Reply by pi0** [reply]
Having a quick look, it seems we are having a shared context issue. When defining a variable like `const state = ref(...)` outside of `<script setup>` it will be created once and shared across all SSR requests! What you can do instead, is use [`useState`](https://v3.nuxtjs.org/docs/usage/state). 

The other thing observed you are using Class-based models. Classes are not supported to be hydrated at the moment. You need to use plain objects in state or at lease implement `toJSON`.

Please check this example: https://stackblitz.com/edit/github-y3vmo9?file=app.vue

Hopefully by avoiding global refs memory leaks should be gone otherwise please ping me when refactored to investigate there are other issues.

**Reply by LeonardoMussato** [reply]
Thank You! I was using the wrong approach to the composables...
I tried your approach and it works, now I just need a day to see if in aws will give me any problems.

Further question: I read the docs but it doesn't explain what is the best practice to change the state, is it possible to change the state of a composable just inside a vue file or also outside? And if so what is the best practice to do it?

PS: Great job for this framework!

**Reply by Rigo-m** [reply]
@LeonardoMussato to have a shared state between vue components you can use `useState` with the same key in both components. Everything needs to be inside the vue setup lifecycle. If you want to have data just inside a vue component you can use a `ref` if it's not something which you want to bring down from SSR, otherwise just use `useState` or `useAsyncData` or `useFetch` depending on your needs providing a *unique* key (so that it doesn't get shared between components.

**Reply by zss1004504439** [reply]
> Having a quick look, it seems we are having a shared context issue. When defining a variable like outside of it will be created once and shared across all SSR requests! What you can do instead, is use [`useState`](https://v3.nuxtjs.org/docs/usage/state).`const state = ref(...)``<script setup>`
> 
> The other thing observed you are using Class-based models. Classes are not supported to be hydrated at the moment. You need to use plain objects in state or at lease implement .`toJSON`
> 
> Please check this example: https://stackblitz.com/edit/github-y3vmo9?file=app.vue
> 
> Hopefully by avoiding global refs memory leaks should be gone otherwise please ping me when refactored to investigate there are other issues.

why "TypeError: product.discount is not a function" ?
![image](https://github.com/user-attachments/assets/df23ce86-5f60-44e5-a9f5-9d8c3e0dc1cd)
reason "Classes are not supported to be hydrated at the moment." ?

**Reply by danielroe** [reply]
You can define your own payload reviver/reducer for custom classes to hydrate/serialize them:

https://nuxt.com/docs/api/composables/use-nuxt-app#payload

**Reply by paulmelero** [reply]
> What you can do instead, is use [`useState`](https://v3.nuxtjs.org/docs/usage/state).

This happened to us with `useState`: We believe that passing an object defined in another module inside `useState` caused a memory leak. In the culprit PR (after bisecting), we started calling a Nuxt composable that used `useState` and this method was being fed with a (small) object defined outside of the composable.

**Comment by psycongaroo** [comment]
So we cannot use ref in composables?

**Reply by danielroe** [reply]
Yes, you can use `ref` in composables 👍.

**Reply by psycongaroo** [reply]
Thank you @danielroe, sorry for wondering.. but I didn't understand the @pi0 's message above, in the repo that Leonardo provided he is using composables in setup().

**Reply by psycongaroo** [reply]
@pi0 @LeonardoMussato @danielroe 

Sorry for tags you guys, can someone explain better if we can use ref or we should use useState?
Thank you so much for your efforts! <3

**Reply by danielroe** [reply]
The issue is not using ref within a composable, but using a ref as a substitute for a global state _outside_ of a composable.

**Reply by psycongaroo** [reply]
> The issue is not using ref within a composable, but using a ref as a substitute for a global state _outside_ of a composable.
@danielroe 

Having a quick look to this repo: https://github.com/LeonardoMussato/nuxt-memory-leak
He's using ref just in setup() not for a global state, if you can give it a look to composable useProducts.

**Reply by danielroe** [reply]
That's a good example - it's being used as a global state: https://github.com/LeonardoMussato/nuxt-memory-leak/blob/master/composables/data.ts

**Reply by psycongaroo** [reply]
Thank you so much!
Now everything's clear.

Goodjob with nuxt3, hope too see a stable version soon. <3
